<?xml version="1.0" encoding="utf-8"?>
<robot name="surgical_tool" xmlns:xacro="http://www.ros.org/wiki/xacro">


  <xacro:include filename="$(find surgical_sim)/urdf/surgical_tool/templates/joint.transmissions.xacro"/>
  <xacro:include filename="$(find surgical_sim)/urdf/surgical_tool/templates/gazebo.joints.xacro"/>

  <xacro:macro name="create_forceps_joints">
    
    <!-- joint 5: tool wrist pitch joint -->
    <joint name="tool_wrist_pitch" type="revolute">
      <parent link="tool_wrist_shaft_link"/> 
      <!-- <parent link="world"/> -->
      <child link="tool_wrist_forceps_link"/>
      <axis xyz="0 0 1"/>
      <origin rpy="-1.5708 -1.5708 ${pi}" xyz="0 0 0"/>
      <limit lower="-${pi/2}" upper="${pi/2}" velocity=".4" effort="1000"/>
    </joint>
    
    <!-- joint 6: Tool wrist yaw joint -->
    <joint name="tool_wrist_yaw" type="revolute">
      <parent link="tool_wrist_forceps_link"/>
      <child link="tool_wrist_forceps_link_shaft"/>
      <axis xyz="0 0 1"/>
      <origin rpy="0.00 -1.5708 -1.5708" xyz="0.0091 0 0"/>
      <limit lower="-${pi/2}" upper="${pi/2}" velocity=".4" effort="1000"/>
    </joint>
    
    <!-- joint 6-1: tool open angle 1 -->
    <joint name="tool_wrist_open_angle" type="revolute">
      <parent link="tool_wrist_forceps_link_shaft"/>
      <child link="tool_wrist_forceps_link_2_left"/>
      <axis xyz="0 0 1"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <limit lower="0" upper="${pi/2}" velocity=".4" effort="1000"/>
    </joint>

  <!-- joint 6-2: tool open angle 2  -->
    <joint name="tool_wrist_open_angle_mimic" type="revolute">
      <parent link="tool_wrist_forceps_link_shaft"/>
      <child link="tool_wrist_forceps_link_2_right"/>
      <axis xyz="0 0 1"/>
      <origin rpy="3.1415926 0 0" xyz="0 0 0"/>
      <limit lower="0" upper="${pi/4}" velocity=".4" effort="1000"/>
      <mimic joint="tool_wrist_open_angle" multiplier="1"/>
    </joint>

         <!-- tool tip -->
    <joint name="tool_tip_" type="fixed">
      <parent link="tool_wrist_forceps_link_shaft"/>
      <child link="tool_tip_link"/>
      <origin rpy="0 ${pi/2} ${pi/2}" xyz="0 0.0202 0"/>
    </joint>
    
    <joint name="tool_tip_ee" type="fixed">
      <parent link="tool_tip_link"/>
      <child link="tool_tip_link_ee"/>
      <origin rpy="${pi} 0 ${-pi/2}" xyz="0 0 0"/>
    </joint>

    <xacro:joint_trans name="tool_wrist_pitch" reduction="1"/>
    <xacro:joint_trans name="tool_wrist_yaw" reduction="1"/>
    <xacro:joint_trans name="tool_wrist_open_angle" reduction="1"/>
    <xacro:joint_trans name="tool_wrist_open_angle_mimic" reduction="1"/>

    <xacro:gazebo_joints name="tool_wrist_pitch"/>
    <xacro:gazebo_joints name="tool_wrist_yaw"/>
    <xacro:gazebo_joints name="tool_wrist_open_angle"/>
    <xacro:gazebo_joints name="tool_wrist_open_angle_mimic"/>

  </xacro:macro>

</robot>


